internal/repository/validator.go

```diff
@@ -7,198 +7,339 @@ import (
+			AppAddress:      appAddress,
+			FirstBlock:      firstBlock,
+			LastBlock:       lastBlock,
+			ClaimHash:       claimHash,
+			TransactionHash: transactionHash,
+			Status:          EpochStatus(status),
+		}
+		results = append(results, epoch)
+		return nil
+	})
+	if err != nil {
+		return nil, fmt.Errorf("GetProcessedEpochs failed: %w\n", err)
+	}
+	return results, nil
 }
 
-func (pg *Database) GetAllOutputsFromProcessedInputs(
-	ctx context.Context,
-	startBlock uint64,
-	endBlock uint64,
-	timeout *time.Duration,
-) ([]Output, error) {
-	ctxTimeout, cancel := context.WithTimeout(ctx, *timeout)
-	defer cancel()
-	for {
-		select {
-		case <-ctxTimeout.Done():
-			return nil, fmt.Errorf("GetAllOutputsFromProcessedInputs timeout") // timeout
-		default:
-			outputs, err := pg.getAllOutputsFromProcessedInputs(ctxTimeout, startBlock, endBlock)
-			if outputs != nil {
-				return outputs, nil
-			}
-
-			if err != nil {
-				return nil, err
-			}
+// GetLastInputOutputHash returns the outputs Merkle tree hash calculated
+// by the Cartesi Machine after it processed the last input in the provided
+// epoch.
+func (pg *Database) GetLastInputOutputHash(ctx context.Context, epoch *Epoch) (*Hash, error) {
+	if epoch == nil {
+		return nil, NilEpoch
+	}
+
+	//Get Epoch from Database
+	epoch, err := pg.GetEpoch(ctx, epoch.Index, epoch.AppAddress)
+	if err != nil {
+		return nil, err
+	}
+
+	//Check Epoch Status
+	switch epoch.Status { //nolint:exhaustive
+	case EpochStatusOpen, EpochStatusClosed:
+		err := fmt.Errorf(
+			"epoch '%d' of app '%v' is still being processed",
+			epoch.Index, epoch.AppAddress,
+		)
+		return nil, err
+	default:
+		break
+	}
+
+	//Get epoch last input
+	query := fmt.Sprintf(`
+	SELECT
+		outputs_hash
+	FROM 
+		input
+	WHERE
+		epoch_id = %d
+	ORDER BY
+		index DESC
+	LIMIT 1
+	`, epoch.Id)
+	var outputHash Hash
+
+	err = pg.db.QueryRow(ctx, query).Scan(&outputHash)
+	if err != nil {
+		if errors.Is(err, pgx.ErrNoRows) {
+			slog.Warn(
+				"no inputs",
+				"service", "repository",
+				"epoch", epoch.Index,
+				"app", epoch.AppAddress,
+			)
+			return nil, nil
 		}
+		return nil, fmt.Errorf("GetLastInputOutputHash failed: %w\n", err)
 	}
+	return &outputHash, nil
 }
 
-func (pg *Database) getAllOutputsFromProcessedInputs(
-	ctx context.Context,
-	startBlock uint64,
-	endBlock uint64,
-) ([]Output, error) {
+// GetPreviousEpoch returns the epoch that ended one block before the start
+// of the current epoch
+func (pg *Database) GetPreviousEpoch(ctx context.Context, currentEpoch *Epoch) (*Epoch, error) {
+	if currentEpoch == nil {
+		return nil, NilEpoch
+	} else if currentEpoch.FirstBlock == 0 {
+		// if this is the first epoch, there is nothing to return
+		return nil, nil
+	}
+
 	query := `
 	SELECT
-		o.id,
-		o.index,
-		o.raw_data,
-		o.input_id,
-		i.status
+		id,
+    	application_address,
+        index,
+        first_block,
+		last_block,
+		claim_hash,
+		transaction_hash,
+		status
```

Comment:
Wrong indentation here as well

Reply:
Done in 59b274d212eeba3100d893bd0df2cd9a02987282